File: problems/1992-find-all-groups-of-farmland.ts
export {}

type Group = [number, number, number, number]

function findFarmland(land: number[][]): Group[] {
  const m = land.length
  const n = m > 0 ? land[0].length : 0
  const groups: Group[] = []

  const visited: boolean[][] = Array.from({ length: m }, () =>
    Array(n).fill(false)
  )

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (visited[i][j] || land[i][j] === 0) {
        continue
      }
      let [r1, c1, r2, c2] = [i, j, i, j]

      // Find bottom edge.
      while (r2 + 1 < m && land[r2 + 1][j] === 1) {
        r2++
      }

      // Find right edge.
      while (c2 + 1 < n && land[i][c2 + 1] === 1) {
        c2++
      }

      groups.push([r1, c1, r2, c2])

      for (let k = r1; k <= r2; k++) {
        for (let l = c1; l <= c2; l++) {
          visited[k][l] = true
        }
      }
    }
  }

  return groups
}

File: problems/1992_find_all_groups_of_farmland.py
class Solution:
    def findFarmland(self, land: list[list[int]]) -> list[list[int]]:
        m = len(land)
        n = len(land[0]) if m else 0
        visited = [[False] * n for _ in range(m)]
        groups: list[list[int]] = []

        for i in range(m):
            for j in range(n):
                if visited[i][j] or land[i][j] == 0:
                    continue
                r1 = i
                c1 = j
                r2 = i
                c2 = j

                # Determine bottom edge.
                while r2 + 1 < m and land[r2 + 1][j] == 1:
                    r2 += 1

                # Determine right edge.
                while c2 + 1 < n and land[i][c2 + 1] == 1:
                    c2 += 1

                groups.append([r1, c1, r2, c2])

                for k in range(r1, r2 + 1):
                    for k2 in range(c1, c2 + 1):
                        visited[k][k2] = True

        return groups


Solution().findFarmland([[1, 0, 0], [0, 1, 1], [0, 1, 1]])
